linux/src/async_socket_linux.c

```diff
@@ -0,0 +1,695 @@
+
+typedef struct ASYNC_SOCKET_TAG
+{
+    int socket_handle;
+    EXECUTION_ENGINE_HANDLE execution_engine;
+
+    volatile_atomic int32_t state;
+    volatile_atomic int32_t pending_api_calls;
+
+    int epoll;
+} ASYNC_SOCKET;
+
+typedef struct ASYNC_SOCKET_OPEN_CONTEXT_TAG
+{
+    ON_ASYNC_SOCKET_OPEN_COMPLETE on_open_complete;
+    void* on_open_complete_context;
+    ASYNC_SOCKET* async_socket;
+} ASYNC_SOCKET_OPEN_CONTEXT;
+
+typedef struct ASYNC_SOCKET_IO_CONTEXT_TAG
+{
+    ASYNC_SOCKET* async_socket;
+    uint32_t total_buffer_bytes;
+    ON_ASYNC_SOCKET_SEND_COMPLETE on_send_complete;
+    void* on_send_complete_context;
+} ASYNC_SOCKET_IO_CONTEXT;
+
+typedef struct ASYNC_SOCKET_RECV_CONTEXT_TAG
+{
+    ASYNC_SOCKET* async_socket;
+    uint32_t total_buffer_bytes;
+    uint32_t total_buffer_count;
+    int socket_handle;
+    ON_ASYNC_SOCKET_RECEIVE_COMPLETE on_receive_complete;
+    void* on_receive_complete_context;
+    ASYNC_SOCKET_BUFFER recv_buffers[];
+} ASYNC_SOCKET_RECV_CONTEXT;
+
+typedef struct ASYNC_SOCKET_SEND_CONTEXT_TAG
+{
+    ASYNC_SOCKET* async_socket;
+    uint32_t total_buffer_bytes;
+    uint32_t total_buffer_count;
+    int socket_handle;
+    ON_ASYNC_SOCKET_SEND_COMPLETE on_send_complete;
+    void* on_send_complete_context;
+    ASYNC_SOCKET_BUFFER buffers[];
+} ASYNC_SOCKET_SEND_CONTEXT;
+
+static int thread_worker_func(void* parameter)
+{
+    (void)parameter;
+
+    // Loop while true
+    do
+    {
+        struct epoll_event events[MAX_EVENTS_NUM];
+        int num_ready = epoll_wait(g_epoll, events, MAX_EVENTS_NUM, EVENTS_TIMEOUT);
+        for(int index = 0; index < num_ready; index++)
+        {
+            if (events[index].events & EPOLLRDHUP)
+            {
+                ASYNC_SOCKET_RECV_CONTEXT* recv_context = events[index].data.ptr;
+                if (recv_context != NULL)
+                {
+                    recv_context->on_receive_complete(recv_context->on_receive_complete_context, ASYNC_SOCKET_RECEIVE_ABANDONED, 0);
+                }
+            }
+            else if (events[index].events & EPOLLIN)
+            {
+                // Receive data from the socket
+                ASYNC_SOCKET_RECEIVE_RESULT receive_result;
+                ASYNC_SOCKET_RECV_CONTEXT* recv_context = events[index].data.ptr;
+
+                if (recv_context != NULL)
+                {
+                    uint32_t index = 0;
+                    ssize_t recv_size;
+                    do
+                    {
+
+                        recv_size = recv(recv_context->socket_handle, recv_context->recv_buffers[index].buffer, recv_context->recv_buffers[index].length, 0);
+                        if (recv_size < 0)
+                        {
+                            if (errno == EAGAIN || errno == EWOULDBLOCK)
+                            {
+                                // No more data to recv
+                                receive_result = ASYNC_SOCKET_RECEIVE_OK;
```

Comment:
Doesn't this actually say that there will be more data?

Reply:
The documentation says that you'll get this if "The socket's file descriptor is marked O_NONBLOCK and no data is waiting to be received", so no data is available, so I think we should return on this ... or should we loop again assuming data is coming on the next iteration?  I'll need to thin about this.